core/src/main/java/org/elasticsearch/common/xcontent/XContentBuilder.java

```diff
@@ -512,29 +525,73 @@ public XContentBuilder rawValue(BytesReference content) throws IOException {
-        if (humanReadable) {
-            field(readableFieldName, timeValue.toString());
-        }
-        field(rawFieldName, timeValue.millis());
-        return this;
+    public XContentBuilder field(String name, TimeValue value) throws IOException {
+        return field(name, value, timeUnit());
     }
 
-    public XContentBuilder timeValueField(String rawFieldName, String readableFieldName, long rawTime) throws IOException {
-        if (humanReadable) {
-            field(readableFieldName, new TimeValue(rawTime).toString());
+    public XContentBuilder field(String name, TimeValue value, TimeUnit timeUnit) throws IOException {
+        if (timeUnit == null) {
+            throw new IllegalArgumentException("Time unit cannot be null");
         }
-        field(rawFieldName, rawTime);
-        return this;
-    }
 
-    public XContentBuilder timeValueField(String rawFieldName, String readableFieldName, long rawTime, TimeUnit timeUnit) throws
-        IOException {
         if (humanReadable) {
-            field(readableFieldName, new TimeValue(rawTime, timeUnit).toString());
+            field(name);
+            value(value);
         }
-        field(rawFieldName, rawTime);
-        return this;
+
+        String suffix;
+        Long raw = null;
+        switch (timeUnit) {
+            case NANOSECONDS:
+                suffix = "_in_nanos";
+                if (value != null) {
+                    raw = value.nanos();
+                }
+                break;
+            case MICROSECONDS:
+                suffix = "_in_micros";
+                if (value != null) {
+                    raw = value.micros();
+                }
+                break;
+            case MILLISECONDS:
+                suffix = "_in_millis";
+                if (value != null) {
+                    raw = value.millis();
+                }
+                break;
+            case SECONDS:
+                suffix = "_in_seconds";
+                if (value != null) {
+                    raw = value.seconds();
+                }
+                break;
+            case MINUTES:
+                suffix = "_in_minutes";
+                if (value != null) {
+                    raw = value.minutes();
+                }
+                break;
+            case HOURS:
+                suffix = "_in_hours";
+                if (value != null) {
+                    raw = value.hours();
+                }
+                break;
+            case DAYS:
+                suffix = "_in_days";
+                if (value != null) {
+                    raw = value.days();
+                }
+                break;
+            default:
+                throw new IllegalArgumentException("Unsupported time unit");
+        }
+        return field(name + suffix).value(raw);
+    }
+
+    public XContentBuilder value(TimeValue value) throws IOException {
```

Comment:
I'm not sure this is ever the right thing to do though - it'll spit out only the user readable version.

Also I think `field(String, Object)` will need so work so it forwards to your new implementations.

I think `value(Object)` will need some work as well. I think maybe it is right to just disallow `TimeValue` there. Make it only ok to use `TimeValue` with `field` variants.